src/types/internal/schema-field.ts

```diff
@@ -0,0 +1,328 @@
+
+  abstract toString(): string;
+
+  abstract normalizeForHash(): string;
+
+  clone(): SchemaField {
+    return SchemaField.create(this);
+  }
+
+  // tslint:disable-next-line: no-any
+  toLiteral(): {} {
+    return {
+      kind: this.kind,
+      annotations: this.annotations,
+      refinement: this.refinement ? this.refinement.toLiteral() : null
+    };
+  }
+
+  equals(other: SchemaField): boolean {
+    // TODO(cypher1): structural check instead of stringification.
+    return this.toString() === other.toString();
+  }
+
+  // TODO(shans): output AtLeastAsSpecific here. This is necessary to support
+  // refinements on nested structures and references.
+  isAtLeastAsSpecificAs(other: SchemaField): boolean {
+    assert(this.kind === other.kind);
+    return this.equals(other);
+  }
+
+  static create(theField: {}|string): SchemaField {
+    let newField = null;
+    // tslint:disable-next-line: no-any
+    const field = theField as any;
+    if (typeof(field) === 'string') {
+      assert(['Text', 'URL', 'Number', 'Boolean', 'Bytes'].includes(field), `non-primitive schema type ${field} need to be defined as a parser production`);
+      newField = new PrimitiveField(field as SchemaPrimitiveTypeValue);
+    } else {
+      switch (field.kind) {
+        case Kind.Primitive:
+          newField = new PrimitiveField(field.type);
+          break;
+        case Kind.KotlinPrimitive:
+          newField = new KotlinPrimitiveField(field.type);
+          break;
+        case Kind.Collection:
+          newField = new CollectionField(SchemaField.create(field.schema));
+          break;
+        case Kind.Reference:
+          newField = new ReferenceField(SchemaField.create(field.schema));
+          break;
+        case Kind.OrderedList:
+          newField = new OrderedListField(SchemaField.create(field.schema));
+          break;
+        case Kind.Inline:
+        case Kind.TypeName:
+          newField = new InlineField(field.model);
+          break;
+        case Kind.Union:
+          newField = new UnionField(field.types.map(type => SchemaField.create(type)));
+          break;
+        case Kind.Tuple:
+          newField = new TupleField(field.types.map(type => SchemaField.create(type)));
+          break;
+        case Kind.Nested:
+          newField = new NestedField(SchemaField.create(field.schema));
+          break;
+        default:
+          throw new Error(`Unsupported schema field ${field.kind}`);
+      }
+    }
+    newField.refinement = field.refinement || null;
+    newField.annotations = field.annotations || [];
+    return newField;
+  }
+
+  // The implementation of fromLiteral creates a cyclic dependency, so it is
+  // separated out. This variable serves the purpose of an abstract static.
+  static fromLiteral: SchemaFieldMethod = null;
+}
+
+export class PrimitiveField extends SchemaField {
+  constructor(public readonly type: SchemaPrimitiveTypeValue) {
+    super(Kind.Primitive);
+    assert(this.type);
```

Comment:
WDYT about having the constructors take the primitives that make sense rather than a generic schema? `create` could be responsible for pulling the components out of the schema (it already sort of is, e.g. it already recurses to produce sub-structure).

This would allow people to construct the various SchemaField types directly, without needing to put a schema AST together first.

Reply:
I'm not sure I understand how it is different from what i currently have. You could already do:
```
const primitive = new PrimitiveField('Text');
const collection = new CollectionField(primitive);
```
etc

Reply:
yeah, you're right. I must have been reading this wrong - I thought you were directly passing ASTs in.